emukit/bayesian_optimization/acquisitions/local_penalization.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+from scipy.stats import norm
+
+from emukit.core.acquisition import Acquisition
+
+
+class LocalPenalization(Acquisition):
+    """
+    Penalization function used in the local penalization method. It is the log of the following function:
+
+    .. math::
+        h(x) = \sum_i{\log\\left(\Phi\\left(\\frac{\|x - x_i\|_{2} - r_i}{s_i}\\right)\\right)}
+
+    where :math:`x_i` are the points already in the batch
+    """
+    def __init__(self, model):
+        """
+        :param model: Model
+        """
+
+        self.x_batch = None
+        self.radius = None
+        self.scale = None
+        self.model = model
+
+    @property
+    def has_gradients(self):
+        return True
+
+    def update_batches(self, x_batch, lipschitz_constant, f_min):
+        """
+        Updates the batches internally and pre-computes the parameters of the penalization function
+        """
+        self.x_batch = x_batch
+        if x_batch is not None:
+            self._compute_parameters(x_batch, lipschitz_constant, f_min)
+        else:
+            self.x_batch = None
+            self.radius = None
+            self.scale = None
+
+    def _compute_parameters(self, x_batch, lipschitz_constant, f_min):
+        """
+        Pre-computes the parameters of a penalization function
+        """
+        mean = self.model.predict(x_batch)[0]
```

Comment:
can you combine this line with the one below. seems like you compute the same stuff twice.

Reply:
Good point, yes I can.